src/anomalib/engine/engine.py

```diff
@@ -0,0 +1,189 @@
+
+    Args:
+        callbacks: Add a callback or list of callbacks.
+    """
+
+    def __init__(
+        self,
+        callbacks: list[Callback] = [],
+        normalizer: NormalizationMethod | DictConfig | Callback | str = NormalizationMethod.MIN_MAX,
+        threshold: BaseThreshold
+        | tuple[BaseThreshold, BaseThreshold]
+        | DictConfig
+        | ListConfig
+        | str = F1AdaptiveThreshold(),
+        task: TaskType = TaskType.SEGMENTATION,
+        image_metrics: list[str] | None = None,
+        pixel_metrics: list[str] | None = None,
+        visualization: DictConfig | None = None,
+        **kwargs,
+    ) -> None:
+        self._cache = _Cache(callbacks=[*callbacks, RichProgressBar()], **kwargs)
+        self.normalizer = normalizer
+        self.threshold = threshold
+        self.task = task
+        self.image_metric_names = image_metrics
+        self.pixel_metric_names = pixel_metrics
+        self.visualization = visualization
+
+        self.lightning_module: AnomalyModule
+        self._trainer: Trainer | None = None
+
+    @property
+    def trainer(self):
+        if not self._trainer:
+            msg = "``self.trainer`` is not assigned yet."
+            raise UnassignedException(msg)
+        return self._trainer
+
+    def _setup_trainer(self, model: AnomalyModule):
+        """Instantiates the trainer based on the model parameters."""
+        if self._cache.needs_override(model) or self._trainer is None:
+            self._cache.update(model)
+            self._trainer = Trainer(**self._cache.params)
+            # Callbacks need to be setup later as they depend on default_root_dir from the trainer
+            self._setup_anomalib_callbacks()
+
+    def _setup_anomalib_callbacks(self):
+        """Setup callbacks for the trainer."""
+
+        _callbacks: list[Callback] = [_PostProcessorCallback()]
+        normalization_callback = get_normalization_callback(self.normalizer)
+        if normalization_callback is not None:
+            _callbacks.append(normalization_callback)
+
+        _callbacks.append(_ThresholdCallback(self.threshold))
+        _callbacks.append(_MetricsCallback(self.task, self.image_metric_names, self.pixel_metric_names))
+
+        if self.visualization is not None:
+            image_save_path = self.visualization.pop("image_save_path", None)
+            if image_save_path is None:
+                image_save_path = self.trainer.default_root_dir + "/images"
+            _callbacks += get_visualization_callbacks(
+                task=self.task, image_save_path=image_save_path, **self.visualization
+            )
+
+        self.trainer.callbacks = _CallbackConnector._reorder_callbacks(self.trainer.callbacks + _callbacks)
+
+    def fit(
+        self,
+        model: AnomalyModule,
+        train_dataloaders: TRAIN_DATALOADERS | LightningDataModule = None,
+        val_dataloaders: EVAL_DATALOADERS = None,
```

Comment:
Is mypy ok with this ?

```suggestion
        train_dataloaders: TRAIN_DATALOADERS | LightningDataModule | None = None,
        val_dataloaders: EVAL_DATALOADERS | None = None,
```